### ISSUE TITLE
Pick existing asset and subset

### ISSUE THREAD

**Issue by mottosso** [getavalon/core#194]
### Goal

Reduce duplicate subsets and increase confidence when creating new versions of existing assets.

### Motivation

At the moment, the asset and subset published to is chosen via the Creator.

![untitled project](https://user-images.githubusercontent.com/2152766/27124052-afd9f3ca-50e7-11e7-818b-80cb7e3fec81.gif)

A warning is produced if the chosen asset does not exist, but so far no indication is made about subset. (see #179)

### Implementation

Provide a drop-down menu alongside the Asset and Subset text fields, listing available assets and subsets of the chosen asset.

This could later then act to fill in any additional values added to the creator dialog, such as tags (#154) and membership of nodes.

**Comment by BigRoy** [comment]
### Picking the asset

How about making the `asset` field read only and adding a "search" button next to it that would pop up the asset widget? For example, popping up something like this:

![set_asset_window](https://user-images.githubusercontent.com/2439881/27550906-a479c3a2-5aa1-11e7-8b9e-c1ae243f00a8.png)

It might be slightly slower for the user than typing directly in the field (if you know the exact name of the asset) but it would be a lot clearer in most scenarios?

### Picking the subset

This could have an autocompletion set up for subsets that might already exist, e.g. as you type it might recommend already existing ones. Then it could be color coded to identify whether you're defining a new subset or it's an already existing one.

Another thing noted about this field is that it might want to have some "default recommendations" as discussed here: https://gitter.im/mindbender-pipeline/Lobby?at=594d005bc101bc4e3a11b43d

This would mean the creator would be able to define _multiple subset defaults_ that it could [recommend an artist to use](https://gitter.im/mindbender-pipeline/Lobby?at=594d00e557a6e9f72eeeb27b). An example of default recommendations can be seen in our (Colorbleed) old pipeline in this image:

![subset_recommended_names](https://files.gitter.im/mindbender-pipeline/Lobby/yHK5/afbeelding.png)

_Also note in this image the extra use of icons and descriptions for the creators, which could be an additional issue of its own._

**Comment by BigRoy** [comment]
### Picking subsets (with recommended defaults)

![detailed](https://user-images.githubusercontent.com/2439881/27551414-99e061a6-5aa3-11e7-874b-6d066e3fc394.png)

Also, here's a mockup over that last image (visually it might be complicated to implement like this, but it's to showcase the overall idea).  The _unused_ subsets would be recommended defaults that are not used yet, the _exists_ ones would be those that were already created.

By default on a new asset those would *all* be grey, since they are all unused. If publishing an asset with a non-default recommended subset it would just show up next time in that list for that asset as one of the already existing ones.

>  Note that for a _Shot_ this list could grow large, because it could have many published characters. As such on the right side under subsets it could show _villain01, villain02, villain03, etc_ all with the "exists" labeling.

**Comment by mottosso** [comment]
All good ideas.

Here's how I would recommend we tackle this; as individual PRs, in this order:

- Phase 1, the user types the name of a subset and a check is made after a second or so, leaving a message at the bottom of the UI like it does for the asset
- Phase 2, autocompletion in the text-box, we can easily do a database query for available members and populate autocompletion with it. There is a [built-in widget](http://doc.qt.io/qt-4.8/qcompleter.html) for just such an occasion.
- Phase 3, a search icon is added to the text field that, once clicked, brings up a more elaborate window in which the user can search and filter his choice more delicately.

Once we've gotten to phase 1, I'd expect us to get more clarity on whether auto-completion would help, and if so get clarification that it is indeed useful and not a wasted effort. At phase 2 I'd imagine it would become clear whether there is a need for more control and what exactly is missing from the existing type + autocomplete functionality.

**Comment by aardschok** [comment]
> Phase 1, the user types the name of a subset and a check is made after a second or so, leaving a message at the bottom of the UI like it does for the asset

What kind of message would you expect when the subset does not exists yet or when it already does?

* `You are creating a new subset` or `New subset 'modelHigh'`
* `You are working on an existing subset` or `Subset already exists` 

**Give the existing subsets?**
I do understand the needto inform the user about the existing subsets, it will let the user know he or she will create a new version of subset `modelHigh` or an entire new subset. Wouldn't it be better to ensure the user can pick the subset if it already exists? With pick I mean use a combo box or list view to bring the options to the user and of course with the feature to create a new one ;)

**Better to know what you got**
> Phase 2, autocompletion in the text-box, we can easily do a database query for available members and populate autocompletion with it. There is a built-in widget for just such an occasion.

If this is going to be the case it might be a better idea to query all subsets of the asset at the launch of the application. Currently the query is done whenever the `name text field` or the `asset text field` is changed but as we work within a context of an asset we can easily do it is the start. 
Here is some pseudo code:

```python
asset = io.find_one(filter={"type": "asset", "name": os.environ["AVALON_ASSET"]})
subsets = io.find(filter={"type": "subset", "parent": asset["_id"]}, projection={"name": 1})
if subsets:
    # subset menu needs to be a fancy textfield / combobox thingy in one
    subsets_menu.addItems([subset["name"] for subset in subsets])
```

**Should we lock down the asset name?**
I have scouted through the code of the Instance Creator to see where the asset name is being fetched and as expected it was through `os.environ["AVALON_ASSET"]`. If this is consistent for the future of the tool (as we work within context I would say YES) than I would opt to lock the text field of the asset name in the UI.

**Keep it simple**
> Phase 3, a search icon is added to the text field that, once clicked, brings up a more elaborate window in which the user can search and filter his choice more delicately.

How elaborate are we talking? The Instance Creator we currently have is simple and precise, I do not see the need for any other features. Could you draw a vision of what you want it to be able to do extra?

## less code == less failure == easy maintenance ##

**Comment by mottosso** [comment]
> Wouldn't it be better to ensure the user can pick the subset if it already exists?

Yes, I think so too.

> Currently the query is done whenever the name text field or the asset text field is changed but as we work within a context of an asset we can easily do it is the start.

It's a bit of an optimisation at this point, as it doesn't change behaviour or function, but I see where you're coming from. An issue with doing it at the start is that there might be a new subset created between the time the user opens the dialog and then hits OK. If we could do a final check upon hitting OK then I'm OK with it.

> than I would opt to lock the text field of the asset name in the UI.

This is actually a feature. It means you are able to publish to one or more assets, from within the context of a single asset. Mindbender requested this as they were working on what could have been variants of a single asset, as multiple assets. Could we make this an dropdown/list/combobox as well?

How about a little arrow next to each textfield, which when the user clicks it pops up a dropdown with available options?

**Comment by mottosso** [comment]
> Could you draw a vision of what you want it to be able to do extra?

From experience, once the number of assets grow into the hundreds, I'd like to facilitate searching by more than a name. I would recommend we reach this stage and experience the problem before we solve it, otherwise it could be classified as overengineering.

**Comment by BigRoy** [comment]
> How about a little arrow next to each textfield, which when the user clicks it pops up a dropdown with available options?

This is similar to what I mentioned here, where the "pick an asset" window was a modal pop-up dialog: https://github.com/getavalon/core/issues/194#issuecomment-311121188

> From experience, once the number of assets grow into the hundreds, I'd like to facilitate searching by more than a name. I would recommend we reach this stage and experience the problem before we solve it, otherwise it could be classified as overengineering.

By reusing the "asset" widget as in my example above that would mean all widgets would automatically allow searching to these same tags and data, making the search features work the same in all areas.

> Should we lock down the asset name?

As @mottosso mentioned, no. This was actually a requested feature to be able to publish to another asset.

**Comment by davidlatwe** [comment]
Base on the implementation proposal, I believe this has been implemented in #283.
